src/libprojectM/Renderer/ShaderEngine.cpp

```diff
@@ -805,237 +749,317 @@ void ShaderEngine::SetupShaderVariables(GLuint program, const Pipeline &pipeline
+    glProgramUniformMatrix4x3fv(program, glGetUniformLocation(program, "rot_rand2"), 1, GL_FALSE, glm::value_ptr(glm::mat4x3(temp_mat[21])));
+    glProgramUniformMatrix4x3fv(program, glGetUniformLocation(program, "rot_rand3"), 1, GL_FALSE, glm::value_ptr(glm::mat4x3(temp_mat[22])));
+    glProgramUniformMatrix4x3fv(program, glGetUniformLocation(program, "rot_rand4"), 1, GL_FALSE, glm::value_ptr(glm::mat4x3(temp_mat[23])));
+
+    // set program uniform "_q[a-h]" values (_qa.x, _qa.y, _qa.z, _qa.w, _qb.x, _qb.y ... ) alias q[1-32]
+    for (int i=0; i < 32; i+=4) {
+        std::string varName = "_q";
+        varName.push_back('a' + i/4);
+        glProgramUniform4f(program, glGetUniformLocation(program, varName.c_str()), pipeline.q[i], pipeline.q[i+1], pipeline.q[i+2], pipeline.q[i+3]);
+    }
+}
 
-			glVertexPointer(2, GL_FLOAT, 0, points);
-			glBlendFunc(GL_ONE,GL_ZERO);
-			glDrawArrays(GL_TRIANGLE_FAN, 0, 4);
+void ShaderEngine::SetupTextures(GLuint program, const Shader &shader)
+{
 
+    uint texNum = 0;
+    for (std::map<std::string, TextureSamplerDesc>::const_iterator iter = shader.textures.begin(); iter
+                    != shader.textures.end(); ++iter)
+    {
+        std::string texName = iter->first;
+        Texture * texture = iter->second.first;
+        Sampler * sampler = iter->second.second;
+        std::string samplerName = "sampler_" + texName;
+
+        // https://www.khronos.org/opengl/wiki/Sampler_(GLSL)#Binding_textures_to_samplers
+        GLint param = glGetUniformLocation(program, samplerName.c_str());
+        if (param < 0) {
+            // unused uniform have been optimized out by glsl compiler
+            continue;
+        }
 
-			glBindTexture(GL_TEXTURE_2D, blur1_tex);
-			glCopyTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, 0, 0, texsize/2, texsize/2);
 
+        glActiveTexture(GL_TEXTURE0 + texNum);
+        glBindTexture(texture->type, texture->texID);
+        glBindSampler(texNum, sampler->samplerID);
 
-		}
+        glUniform1i(param, texNum);
 
-		if (blur2_enabled)
-		{
+        std::string texsizeName = "texsize_" + texName;
+        GLint textSizeParam = glGetUniformLocation(program, texsizeName.c_str());
+        if (param >= 0) {
+            glProgramUniform4f(program, textSizeParam, texture->width, texture->height,
+                               1 / (float) texture->width, 1 / (float) texture->height);
+        } else {
+            std::cerr << "invalid texsize name " << texsizeName << std::endl;
+            return;
+        }
 
+        texNum++;
+    }
+}
 
-			float points[4][2] =
-			{
-			{ 0, 0.25 },
-			{ 0, 0 },
-			{ 0.25, 0 },
-			{ 0.25, 0.25 } };
 
-			glVertexPointer(2, GL_FLOAT, 0, points);
-			glBlendFunc(GL_ONE,GL_ZERO);
-			glDrawArrays(GL_TRIANGLE_FAN, 0, 4);
+void ShaderEngine::RenderBlurTextures(const Pipeline &pipeline, const PipelineContext &pipelineContext)
```

Comment:
maybe the blur stuff should go in a separate file. this file is getting waaayyy too big